learning4e.py

```diff
@@ -453,139 +464,216 @@ def information_content(values):
     return sum(-p * np.log2(p) for p in probabilities)
 
 
-def DecisionListLearner(dataset):
+class DecisionListLearner(Learner):
     """
     [Figure 18.11]
     A decision list implemented as a list of (test, value) pairs.
     """
 
-    def decision_list_learning(examples):
+    def __init__(self, dataset):
+        self.predict.decision_list = self.decision_list_learning(set(dataset.examples))
+
+    def decision_list_learning(self, examples):
         if not examples:
             return [(True, False)]
-        t, o, examples_t = find_examples(examples)
+        t, o, examples_t = self.find_examples(examples)
         if not t:
             raise Exception
-        return [(t, o)] + decision_list_learning(examples - examples_t)
+        return [(t, o)] + self.decision_list_learning(examples - examples_t)
 
-    def find_examples(examples):
+    def find_examples(self, examples):
         """
         Find a set of examples that all have the same outcome under
         some test. Return a tuple of the test, outcome, and examples.
         """
         raise NotImplementedError
 
-    def passes(example, test):
+    def passes(self, example, test):
         """Does the example pass the test?"""
         raise NotImplementedError
 
-    def predict(example):
+    def predict(self, example):
         """Predict the outcome for the first passing test."""
-        for test, outcome in predict.decision_list:
-            if passes(example, test):
+        for test, outcome in self.predict.decision_list:
+            if self.passes(example, test):
                 return outcome
 
-    predict.decision_list = decision_list_learning(set(dataset.examples))
-
-    return predict
 
-
-def NearestNeighborLearner(dataset, k=1):
+class NearestNeighborLearner(Learner):
     """k-NearestNeighbor: the k nearest neighbors vote."""
 
-    def predict(example):
+    def __init__(self, dataset, k=1):
+        self.dataset = dataset
+        self.k = k
+
+    def predict(self, example):
         """Find the k closest items, and have them vote for the best."""
-        best = heapq.nsmallest(k, ((dataset.distance(e, example), e) for e in dataset.examples))
-        return mode(e[dataset.target] for (d, e) in best)
+        best = heapq.nsmallest(self.k, ((self.dataset.distance(e, example), e) for e in self.dataset.examples))
+        return mode(e[self.dataset.target] for (d, e) in best)
 
-    return predict
 
+class LossFunction:
```

Comment:
Why do loss functions need to be a class? I think the code would be much simpler if we kept them as functions.

Reply:
I'm sorry, this class is necessary to fix the X and y matrices before and then pass the methods to Scipy minimization method! Anyway, I think it is right to have classes for the loss functions as for the activation functions that expose two methods, one that returns the cost of the function and the other the derivative (or, in this case, the Jacobian).